### PR TITLE
Made apple verification compatible with in_app format (#10)

### PR DESCRIPTION
**Pull request by sambauwens**

### CODE REVIEW

File: lib/apple/index.js [hunk 1]
```diff
@@ -37,13 +37,30 @@ function parseResult(result) {
...
-		productId: receipt.product_id
+		transactionId: getReceiptFieldValue(receipt, "transaction_id"),
+		productId: getReceiptFieldValue(receipt, "product_id")
 		/* jshint camelcase:true */
 	};
 }
 
 
+function getReceiptFieldValue(receipt, field)
+{
```
Comment (ronkorving) [inline lib/apple/index.js]: Please try to respect the code style of the project (curly braces at the end of the line).

File: lib/apple/index.js [hunk 2]
```diff
@@ -37,13 +37,30 @@ function parseResult(result) {
 	return {
 		receipt: receipt,
 		/* jshint camelcase:false */
-		transactionId: receipt.transaction_id,
-		productId: receipt.product_id
+		transactionId: getReceiptFieldValue(receipt, "transaction_id"),
+		productId: getReceiptFieldValue(receipt, "product_id")
```
Comment (ronkorving) [inline lib/apple/index.js]: All strings should be single quoted in this project.

File: lib/apple/index.js [hunk 3]
```diff
@@ -37,13 +37,30 @@ function parseResult(result) {
 	return {
 		receipt: receipt,
 		/* jshint camelcase:false */
```
Comment (ronkorving) [inline lib/apple/index.js]: No more need for this then :) Can you remove it?

File: lib/apple/index.js [hunk 4]
```diff
@@ -37,13 +37,30 @@ function parseResult(result) {
...
 
+function getReceiptFieldValue(receipt, field)
+{
+    var fieldValue = null;
+
+    if (receipt.hasOwnProperty(field))
+    {
+        fieldValue = receipt[field];
+    }
+    else if (receipt.hasOwnProperty("in_app") && receipt.in_app[0].hasOwnProperty(field))
```
Comment (ronkorving) [inline lib/apple/index.js]: You're assuming in_app[0] is defined. Rewriting this to `if (receipt.in_app && receipt.in_app[0] && receipt.in_app[0].hasOwnProperty(field))` would do the trick.

Comment (ronkorving) [inline lib/apple/index.js]: For that way of referring to in_app however, you will have to add `/* jshint camelcase:false */` in this function (see above)

**Comment by ronkorving** [comment]
Please make sure this PR lints and passes all tests by running `make test`.

**Comment by ronkorving** [comment]
@sambauwens can we work on this PR and try to get it in?

**Comment by sambauwens** [comment]
Yes, sorry I didn't give you a response quicker.

**Comment by tyorikan** [comment]
@sambauwens Would you merge this PR? I'd like to check `bundle_id` of receipt.

**Comment by KLVTZ** [comment]
@ronkorving how's this looking for a merge?

**Comment by ronkorving** [comment]
I need some help to understand something. Why is it that some people end up with a receipt that has this `in_app` property, and other people (like myself) don't? It makes it very hard to test and ensure code remains stable. Is the origin of this 2 different iOS APIs being used?

Also, in the case of an `in_app` property being there, would it make sense to just reduce the receipt data to `receipt.in_app[0]`?

That is:

```js
if (receipt.in_app) { receipt = receipt.in_app[0]; }
```

**Comment by KLVTZ** [comment]
@ronkorving From my understanding the apple receipt can represent different types of purchases. For example, your application may support a one-time purchase for a digital good. That receipt would have a record of the purchase but nothing more. Our application supports monthly and yearly subscription access. As such, the receipt auto-renews and a new subscription is available. That is represented in both the `in_app` field as well as the `latest_receipt_info`. Granted, the documentation on this is not verbose. But from my experience this is more or less what I've noticed.

I don't think we should provide or reduce the output of the receipt in favor of one property or another. We should be a mediator in providing as much details from apple to the client as possible. That's why with #24 we attempt to provide both the latest receipt info (if available) as well as supporting access to `in_app` if available.

**Comment by ronkorving** [comment]
So the difference has everything to do with subscriptions? That would make sense, as I've never tried subscriptions. Glad to have that sorted out. I like the solution in #24 👍

**Comment by KLVTZ** [comment]
@ronkorving I agree, its very confusing at first. Wish Apple added more verbose documentation to receipt validation. Internally we're starting to investigate different flows and how receipts are affected when a user cancels their subscription and signs-up the next month. Out of scope of this project, but just wanted to highlight how Apple could help everyone with better documentation. I think this can be closed by #24

**Comment by ronkorving** [comment]
Yup 👍

**Pull request closed without merging**